modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/management_bar/components/filters/DynamicFilter.js

```diff
@@ -0,0 +1,281 @@
+import {fetch, openToast} from 'frontend-js-web';
+import PropTypes from 'prop-types';
+import React, {useEffect, useState} from 'react';
+
+const getSelectedItemsLabel = ({selectedData}) => {
+	return selectedData.map((item) => item.label).join(', ');
+};
+
+const getOdataString = ({id, selectedData}) => {
+	return `${id} in (${selectedData
+		.map((item) => `'${item.value}'`)
+		.join(',')})`;
+};
+
+async function getPicklist(erc) {
+	const response = await fetch(
+		`/o/headless-admin-list-type/v1.0/list-type-definitions/by-external-reference-code/${erc}`
+	);
+
+	if (!response.ok) {
+		openToast({
+			message: Liferay.Language.get('your-request-failed-to-complete'),
+			type: 'danger',
+		});
+
+		return [];
+	}
+
+	return await response.json();
+}
+
+const DynamicFilter = ({
+	id,
+	listTypeDefinitionId,
+	multiple,
+	preselectedValues,
+	selectedData,
+	setFilter,
+}) => {
+	const [selectedItems, setSelectedItems] = useState(
+		selectedData ? selectedData : preselectedValues
+	);
+	const [sourceItems, setSourceItems] = useState([]);
+	const [availableItems, setAvailableItems] = useState([]);
+
+	useEffect(() => {
+		getPicklist(listTypeDefinitionId).then((picklist) => {
+			const {listTypeEntries} = picklist;
+
+			const availableItems = listTypeEntries.map((entry) => ({
+				label:
+					entry.name_i18n[Liferay.ThemeDisplay.getLanguageId()] ||
+					entry.name,
+				value: String(entry.key),
+			}));
+
+			setAvailableItems(availableItems);
+		});
+	}, [listTypeDefinitionId]);
+
+	useEffect(() => {
+		if (multiple) {
+			const selectedValues = selectedItems.map((item) => item.value);
+
+			setSourceItems(
+				availableItems.filter(
+					(item) => !selectedValues.includes(item.value)
+				)
+			);
+		}
+		else {
+			setSourceItems(selectedItems.length ? [] : availableItems);
+		}
+	}, [availableItems, multiple, selectedItems]);
+
+	return (
+		<>
+			<ClayDropDown.Caption>
+				<ClayMultiSelect
+					allowsCustomLabel={false}
+					items={selectedItems}
+					loadingState={4}
+					onItemsChange={(items) => {
+						const presentValues = {};
+
+						const selectedItems = items.filter((item) => {
+							if (presentValues[item.value]) {
+								return false;
+							}
+
+							presentValues[item.value] = true;
+
+							return true;
+						});
+
+						setSelectedItems(selectedItems);
+					}}
+					sourceItems={sourceItems}
+					value=""
+				/>
+			</ClayDropDown.Caption>
+			<ClayDropDown.Divider />
+			<ClayDropDown.Caption>
+				<ClayButton
+					onClick={() => {
+						if (selectedItems.length) {
+							const selectedData = selectedItems;
+
+							setFilter({
+								active: true,
+								id,
+								odataFilterString: getOdataString({
+									id,
+									selectedData,
+								}),
+								selectedData,
+								selectedItemsLabel: getSelectedItemsLabel({
+									selectedData,
+								}),
+							});
+						}
+						else {
+							setFilter({
+								active: false,
+								id,
+								selectedData: [],
```

Comment:
When you set this, are you expecting that the prop `selectedData` changes? Meaning, `setFilter`

Reply:
I had to add that because the filters UI framework saves the `selectedData` even if the filter is disabled (which makes sense) so, if I don't reset it, it appears prepopulated the next time I open the filter (after having removed it).

In any case, I think there are some minor issues with the filters UI framework state, because playing with the filters: configuring them, editing, then clearing from the edition or from the list header, etc. has some corner cases which I'm not quite sure is what we really want.

However, take my comment with a pinch of salt because there's no documentation saying how it should behave so it's just my opinion about how it works vs how it should work.